osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,774 @@
+    }
+
+    if (message != nullptr) {
+      message = sd_bus_message_unref(message);
+    }
+
+    return Status(1, e.what());
+  }
+}
+
+Status getInactiveSystemdUnitList(std::vector<SystemdUnitInfo>& unit_list,
+                                  sd_bus* bus) {
+  unit_list.clear();
+
+  sd_bus_message* message = nullptr;
+  sd_bus_message* reply = nullptr;
+  sd_bus_error bus_error = SD_BUS_ERROR_NULL;
+
+  try {
+    if (sd_bus_call_method(bus,
+                           "org.freedesktop.systemd1",
+                           "/org/freedesktop/systemd1",
+                           "org.freedesktop.systemd1.Manager",
+                           "ListUnitFiles",
+                           &bus_error,
+                           &reply,
+                           nullptr) < 0) {
+      std::string error_message;
+      if (bus_error.message != nullptr) {
+        error_message = bus_error.message;
+      } else {
+        error_message = "Failed to call the remote method";
+      }
+
+      throw std::runtime_error(error_message);
+    }
+
+    if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "(ss)") < 0) {
+      throw std::runtime_error("Failed to enter the data container");
+    }
+
+    while (true) {
+      const char* unit_path = nullptr;
+      const char* unit_state = nullptr;
+
+      auto error = sd_bus_message_read(reply, "(ss)", &unit_path, &unit_state);
+      if (error == 0) {
+        break;
+      } else if (error < 0) {
+        throw std::runtime_error("Failed to parse the unit information");
+      }
+
+      if (std::strcmp(unit_state, "disabled") != 0) {
+        continue;
+      }
+
+      SystemdUnitInfo info = {};
+      info.unit_path = unit_path;
+      info.active_state = unit_state;
+
+      unit_list.push_back(info);
+    }
+
+    if (sd_bus_message_exit_container(reply) < 0) {
+      throw std::runtime_error("Failed to exit the data container");
+    }
+
+    reply = sd_bus_message_unref(reply);
+    message = sd_bus_message_unref(message);
+
+    if (unit_list.empty()) {
+      return Status(1, "No services returned by the manager!");
+    }
+
+    return Status(0, "OK");
+
+  } catch (const std::exception& e) {
+    if (reply != nullptr) {
+      reply = sd_bus_message_unref(reply);
+    }
+
+    if (message != nullptr) {
+      message = sd_bus_message_unref(message);
+    }
+
+    return Status(1, e.what());
+  }
+}
+
+Status getSystemdUnitList(std::vector<SystemdUnitInfo>& unit_list,
+                          sd_bus* bus) {
+  unit_list.clear();
+
+  std::vector<SystemdUnitInfo> loaded_unit_list;
+  auto status = getLoadedSystemdUnitList(loaded_unit_list, bus);
+  if (!status.ok()) {
+    return status;
+  }
+
+  std::vector<SystemdUnitInfo> inactive_unit_list;
+  status = getInactiveSystemdUnitList(inactive_unit_list, bus);
+  if (!status.ok()) {
+    return status;
+  }
+
+  unit_list = std::move(loaded_unit_list);
+  loaded_unit_list.clear();
+
+  // clang-format off
+  unit_list.insert(
```

Comment:
Is not it waste of time to create 2 extra vectors and copy things around?

Reply:
Yes, I think it would be good. Maybe, changing the get to add/append will not be that confusing.